crates/bevy_geometry/src/lib.rs

```diff
@@ -1,28 +1,183 @@
+}
 
-#[derive(Copy, Clone, PartialEq, Debug)]
+#[derive(Debug, Clone)]
+pub enum PrimitiveError {
+    MinGreaterThanMax,
+    NonPositiveExtents,
+}
+impl Error for PrimitiveError {}
+impl fmt::Display for PrimitiveError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            PrimitiveError::MinGreaterThanMax => {
+                write!(f, "AxisAlignedBox minimums must be smaller than maximums")
+            }
+            PrimitiveError::NonPositiveExtents => {
+                write!(f, "AxisAlignedBox extents must be greater than zero")
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Debug, Reflect)]
 pub struct Sphere {
     pub origin: Vec3,
     pub radius: f32,
 }
+impl Primitive3d for Sphere {}
 
-#[derive(Copy, Clone, PartialEq, Debug)]
-pub struct Box {
-    pub maximums: Vec3,
-    pub minimums: Vec3,
+/// An oriented box, unlike an axis aligned box, can be rotated and is not constrained to match the
+/// orientation of the coordinate system it is defined in. Internally, this is represented as an
+/// axis aligned box with some rotation ([Quat]) applied.
+#[derive(Copy, Clone, PartialEq, Debug, Reflect)]
+pub struct OrientedBox {
+    pub aab: AxisAlignedBox,
     pub orientation: Quat,
 }
+impl Primitive3d for OrientedBox {}
 
-#[derive(Copy, Clone, PartialEq, Debug)]
+/// An axis aligned box is a box whose axes lie in the x/y/z directions of the coordinate system
+/// the box is defined in.
+#[derive(Copy, Clone, PartialEq, Debug, Reflect)]
 pub struct AxisAlignedBox {
-    pub maximums: Vec3,
-    pub minimums: Vec3,
+    minimums: Vec3,
+    maximums: Vec3,
 }
+impl Primitive3d for AxisAlignedBox {}
+impl AxisAlignedBox {
+    pub fn from_min_max(minimums: Vec3, maximums: Vec3) -> Result<AxisAlignedBox, PrimitiveError> {
+        if (maximums - minimums).min_element() > 0.0 {
+            Ok(AxisAlignedBox { minimums, maximums })
+        } else {
+            Err(PrimitiveError::MinGreaterThanMax)
+        }
+    }
+    pub fn from_extents_origin(
+        extents: Vec3,
+        origin: Vec3,
+    ) -> Result<AxisAlignedBox, PrimitiveError> {
+        if extents.min_element() > 0.0 {
+            Ok(AxisAlignedBox {
+                minimums: origin,
+                maximums: extents + origin,
+            })
+        } else {
+            Err(PrimitiveError::NonPositiveExtents)
+        }
+    }
+}
+
+/// A frustum is a truncated pyramid that is used to represent the "volume" of world space that is
+/// visible to the camera.
+#[derive(Copy, Clone, PartialEq, Debug, Reflect)]
+#[reflect_value(PartialEq)]
+pub struct Frustum {
+    planes: [Plane; 6],
+}
+impl Primitive3d for Frustum {}
+impl Frustum {
+    pub fn from_camera_properties(
+        &self,
+        camera_position: Mat4,
+        projection_matrix: Mat4,
+    ) -> Frustum {
+        let ndc_to_world: Mat4 = camera_position * projection_matrix.inverse();
+        // Near/Far, Top/Bottom, Left/Right
```

Comment:
I don't see a way to do this in a loop yet, without making the logic harder to follow. Everything is "unrolled" and named so it's easier to follow which points are being used to take cross products and compute normals. I could probably make the code shorter and less repetitive with a well-constructed loop, but I worry it will be harder to follow and debug.